src/patchset.py

```diff
@@ -0,0 +1,250 @@
+#!/usr/bin/env python3
+#
+# SPDX-License-Identifier: LGPL-2.1-or-later
+#
+# Copyright (C) 2022 Collabora Limited
+# Author: Nikolay Yurin <yurinnick@meta.com>
+
+import os
+import sys
+import json
+import requests
+import time
+import tempfile
+import hashlib
+from datetime import datetime, timedelta
+from urllib.parse import urlparse
+from urllib.request import urlopen
+
+import kernelci
+import kernelci.build
+import kernelci.config
+from kernelci.legacy.cli import Args, Command, parse_opts
+import kernelci.storage
+
+from tarball import Tarball
+
+# FIXME: make patchset service configuration option
+ALLOWED_DOMAINS = {"patchwork.kernel.org"}
+PATCHSET_SHORT_HASH_LEN = 13
+
+
+class Patchset(Tarball):
+    TAR_CREATE_CMD = """\
+set -e
+cd {target_dir}/
+tar --create --transform "s/^/{prefix}\\//" * | gzip > {tarball_path}
+"""
+
+    APPLY_PATCH_SHELL_CMD = """\
+set -e
+cd {kdir}
+patch -p1 < {patch_file}
+"""
+
+    # FIXME: I really don't have a good idea what I'm doing here
```

Comment:
I definitely need some help here, I am not confident enough to say which part we should leave and which we should redact before hashing.

Reply:
I think we've discussed this before.  If I remember correctly, and it's probably in the open hours notes, the consensus was to take all the contents of what `git format-patch` would produce including the commit details and author etc. as well as the base Git SHA1.  This basically would be the same as doing actual Git commits, as the SHA1 revisions include all the meta-data.  That way, you would only get the same patchset hash if and only if the exact same patch files were used with the same base Git revision.  So you know for sure that results would only be collated with other matching revisions, if another user submitted the same patchset with the same base commit.

Reply:
And we also said that an alternative which could be done as a follow-up but would be a bit more intricate, would be to only compute the hash with the diff and the commit message but leave out the rest of the meta-data and base Git commit.  Then clients would need to rely on both the base Git commit in `revision.commit` and `revision.patchset` as a pair of values to identify patchsets.  A same patchset applied on a different base commit would have the same `revision.patchset` hash but obviously a different `revision.commit` value.  The advantage here is to be able to compare same identical patchsets but applied on different base Git revisions, say once on a subsystem maintainer's branch and once on linux-next, as they will have the same `revision.patchset` regardless of how the patch files were created (different commit time etc.) as long as the diff and commit messages are the same.  So we could collate results with a bit more flexibility, at the expense of always checking the two fields.  Well we probably also need to deal with the two fields anyway when using the approach in my previous comment so maybe this is actually the way to go :p

I might be available on Wednesday for the Open Hours to discuss this one more time, maybe one last time though.